### ISSUE TITLE
Run beforeTest only before last element in nested test structure

### ISSUE THREAD

**Issue by LimmaPaulus** [kotest/kotest#882]
Hi!
It seems that beforeTest is currently ran before each "level" of test, at least when using FreeSpec. Would it be possible to run it only before actual test invocation i.e. last element in nested scopes. Consider this example:
```
class Tests : FreeSpec({
  "functionX"-{
      // unnecessary beforeTest-run here
      "when parameter is T"-{
          // unnecessary beforeTest-run here
          "should return value Y"{
              // beforeTest is needed here
              result = functionX(T)
              ...
          }
      }
      "when parameter is K"-{
          // unnecessary beforeTest-run here
          "should return value Z"{
              // beforeTest is needed here
              result = functionX(K)
              ...
          }
      }
  }
})
```

I do some relatively heavy database initializations before my integration tests, and current way to execute beforeTests is slowing testing down quite miserably.

Current behavior may definitely be intuitive and correct for many users, but my test style would need different execution order. Would it be possible to introduce for example new listener function for this scenario?

**Comment by sksamuel** [comment]
See Isolation Mode in the docs.

**Comment by LimmaPaulus** [comment]
Thanks for the suggestion, but I tried with each IsolationMode option and each of them is running beforeTest before each leaf but before each inner context.

But actually I reconsidered my needs, and found that using InstancePerLeaf as IsolationMode and beforeSpec instead of beforeTest is suitable for me. This does not fulfill the example I provided, but now the database is reset before each "test execution path" which is actually perfect for my use case.

So at least for me this issue is not needed any longer. But as I think that my original problem still exists, I let you choose to close or keep the issue :).
